File: Python/ByRound/2008/2008_B_Square_or_Not.py
# Problem: CF 2008 B - Square or Not
# https://codeforces.com/contest/2008/problem/B

# Problem: Check if a binary string could represent a square beautiful matrix.
# A beautiful matrix has 1s on edges and 0s inside.
# Algorithm:
# - For each test case, determine if the string can form a square matrix (r = c)
# - The matrix must have 1s on all edges and 0s in inner positions
# - Key insight: if a square matrix of size n×n exists, then we have:
#   - First row: all 1s
#   - Last row: all 1s
#   - Inner rows: start with 1, end with 1, middle is all 0s (since it's beautiful)
#   - So check if it's possible to partition the string into n rows of length n
#   - And all rows except first and last must start and end with 1 and have all 0s in between
# Time Complexity: O(n) for each test case where n is length of string
# Space Complexity: O(1)

for s in [*open(0)][2::2]:  # Read strings from input, skipping first line (test count)
    print("YNeos"[(s.find("0") - 1) ** 2 != len(s) - 1 :: 2])


# https://github.com/VaHiX/codeForces/

File: CPP/ByTier/B/2008_B_Square_or_Not.cpp
// Problem: CF 2008 B - Square or Not
// https://codeforces.com/contest/2008/problem/B

/*
B. Square or Not
Algorithm: Preprocessing all perfect squares up to 1000*1000 to quickly check if a given string length could correspond to a square matrix.
           Then for each input string, we validate:
           - If the string length is not a perfect square, it's impossible to form a valid square matrix -> "No"
           - Otherwise, we try to build a square matrix of that side length using string characters,
             and verify that:
             - The edges are all '1' and internal cells are all '0'
           Time Complexity: O(n) for preprocessing + O(sqrt(n)) for each query
           Space Complexity: O(1000^2) = O(10^6)

Input Format:
- First line: number of test cases t
- For each test case:
  - Line 1: integer n (length of string)
  - Line 2: binary string s of length n

Output Format:
For each test case, print "Yes" if the string could have been formed from a square beautiful matrix, else "No".
*/

#include <stdio.h>
#include <iostream>
#include <map>
#include <string>

int main() {
  std::map<long, long> msqrt;
  // Precompute all perfect squares up to 1000*1000 and map them to their square root
  for (long p = 1; p <= 1000; p++) {
    msqrt[p * p] = p;
  }
  long t;
  scanf("%ld", &t);
  while (t--) {
    long n;
    scanf("%ld", &n);
    std::string s;
    std::cin >> s;
    // Check if the string length is a perfect square
    if (!msqrt.count(s.size())) {
      puts("No");
      continue;
    }
    long sz = msqrt[s.size()];
    bool possible(true);
    // Iterate through each cell in the matrix, and validate edge vs interior condition
    for (long row = 0; possible && row < sz; row++) {
      for (long col = 0; possible && col < sz; col++) {
        bool digit = (s[row * sz + col] == '1');
        bool edge = (row == 0 || row == sz - 1 || col == 0 || col == sz - 1);
        // For a valid beautiful matrix, an edge should be '1', and an interior should be '0'
        if (edge ^ digit) {
          possible = false;
        }
      }
    }
    puts(possible ? "Yes" : "No");
  }
}


// https://github.com/VaHiX/codeForces/